src/bp/core/services/bot-monitoring-service.ts

```diff
@@ -0,0 +1,116 @@
+import 'bluebird-global'
+import { Logger, LoggerEntry } from 'botpress/sdk'
+import { ConfigProvider } from 'core/config/config-loader'
+import { LogsRepository } from 'core/repositories/logs'
+import { TYPES } from 'core/types'
+import { inject, injectable, tagged } from 'inversify'
+import { Redis } from 'ioredis'
+import _ from 'lodash'
+import moment from 'moment'
+import ms from 'ms'
+import Redlock from 'redlock'
+
+import { JobService } from './job-service'
+
+@injectable()
+export class BotMonitoringService {
+  private _redisClient?: Redis
+  private _redlock!: Redlock
+  private _keyMonitorLock = 'bot_monitor_lock'
+  private _lock
+
+  private _intervalRef: NodeJS.Timeout | undefined
+  private _currentPromise
+  private _lastBotMonitorCheckTs?: Date
+  private _interval?: number
+
+  public onBotError?: (botId: string, events: LoggerEntry[]) => Promise<void>
+
+  constructor(
+    @inject(TYPES.Logger)
+    @tagged('name', 'BotMonitoring')
+    protected logger: Logger,
+    @inject(TYPES.ConfigProvider) private configProvider: ConfigProvider,
+    @inject(TYPES.LogsRepository) private logsRepo: LogsRepository,
+    @inject(TYPES.JobService) private jobService: JobService
+  ) {}
+
+  async start() {
+    const config = await this.configProvider.getBotpressConfig()
+    if (!config?.botMonitoring?.enabled) {
+      return
+    }
+
+    this._redisClient = this.jobService.getRedisClient()
+
+    if (this._redisClient) {
+      this._redlock = new Redlock([this._redisClient], {
+        driftFactor: 0.01,
+        retryCount: 0,
+        retryDelay: 200,
+        retryJitter: 200
+      })
+    }
+
+    if (config.botMonitoring.interval && this.onBotError) {
+      this._interval = ms(config.botMonitoring.interval)
+      this._intervalRef = setInterval(() => this._checkBotMonitor(), this._interval)
+    }
+  }
+
+  stop() {
+    if (this._intervalRef) {
+      clearInterval(this._intervalRef)
+      this._intervalRef = undefined
+    }
+  }
+
+  private _isLockAcquired = async () => {
+    if (!this._redlock) {
+      return true
+    }
+
+    try {
+      const lockDuration = this._interval! + 5000
+      if (this._lock) {
+        await this._lock.extend(lockDuration)
```

Comment:
I feel like this is is not intuitive (it doesn't aligne with the method name),  everytime you check if the lock is acquired and that it is you extend its duration.

I would expect a simple check with no side effects

Reply:
I'm using the same implementation we have in alerting

Reply:
That doesn't make it any more intuitive hehe

Reply:
I'm open to suggestions, I couldn't think of a better naming at the moment. if(acquireLockAndExtend) ? obtainLock? refreshLock?